src/utils/word-api.ts

```diff
@@ -0,0 +1,205 @@
+/* eslint-disable no-console */
+import { CitationResult, StatefulCitation } from "citeproc";
+
+class WordApi {
+  static async insertEmptyContentControl(): Promise<unknown> {
+    return Word.run((context: Word.RequestContext) => {
+      const getSelection = context.document.getSelection();
+      const contentControl = getSelection.insertContentControl();
+      contentControl.tag = "JABREF-CITATION-NEW";
+      contentControl.appearance = "BoundingBox";
+      return context.sync();
+    }).catch((error) => {
+      console.log(`Error: ${JSON.stringify(error)}`);
+      if (error instanceof OfficeExtension.Error) {
+        console.log(`Debug info: ${JSON.stringify(error.debugInfo)}`);
+      }
+    });
+  }
+
+  async setCitations(
+    data: Array<CitationResult>,
+    citationByIndex: Array<StatefulCitation>
+  ): Promise<unknown> {
+    return Word.run(async (context: Word.RequestContext) => {
+      const citationContentControls = context.document.contentControls;
+      context.load(citationContentControls, "tag, length");
+      return context.sync().then(() => {
+        data.forEach((res: CitationResult) => {
+          const position = res[0];
+          const tag = this.getCitationTagForPosition(citationByIndex, position);
+          const citationContentControl =
+            this.getCitationContentControlAtPosition(
+              citationContentControls,
+              position
+            );
+          if (!citationContentControl) {
+            return;
+          }
+          const currentCitationTag = citationContentControl.tag;
+          if (
+            currentCitationTag === "JABREF-CITATION-NEW" ||
+            currentCitationTag !== tag
+          ) {
+            citationContentControl.tag = tag;
+          }
+          citationContentControl.insertText(res[1], "Replace");
+        });
+        return context.sync();
+      });
+    }).catch((error) => {
+      console.log(`Error: ${JSON.stringify(error)}`);
+      if (error instanceof OfficeExtension.Error) {
+        console.log(`Debug info: ${JSON.stringify(error.debugInfo)}`);
+      }
+    });
+  }
+
+  getCitationTagForPosition = (
+    citationByIndex: Array<StatefulCitation>,
+    position: number
+  ): string => {
+    return `JABREF-CITATION-${JSON.stringify(citationByIndex[position])}`;
+  };
+
+  getCitationContentControlAtPosition = (
+    contentControls: Word.ContentControlCollection,
+    position: number
+  ): Word.ContentControl => {
+    let pos = 0;
+    let contentControl: Word.ContentControl = null;
+    for (let i = 0, ilen = contentControls.items.length; i < ilen; i += 1) {
```

Comment:
why not foreach? Or or something like `contentControls.filter(tag.includes(...))[position]`?